README.md
# simple-chat

Big thanks to [Beej's Guide to Networking Programming] for help and inspiration.  Buy his book!

In particular, see his [section on the `select` system call and example code].

> Why use `select`, you may be thinking to yourself?  After all, it only supports a limited number of connections (1024).
>
> Well, I don't think I even know that many people in the whole, wide world.  It's safe to say that even that limited number won't be on `simple-chat` all at the same time.

## Installation

`make build`

Only tested on Linux.  It uses [`select`], so it *should* be portable :)

## Examples

Start the server and connect via `telnet` or [`netcat`].  By default it binds to port 3333, but this can be changed by specifying a different port number.

Start the server:

```
./simple_chat 1992
Simple chat server started on port 1992...
Clients can connect using telnet, i.e., `telnet 127.0.0.1 1992`
```

Connect a client:

```
telnet 127.0.0.1 1992
```

Or:

```
nc 127.0.0.1 1992
```

## License

[GPLv3](COPYING)

## Author

Benjamin Toll

[Beej's Guide to Networking Programming]: https://beej.us/guide/bgnet/
[section on the `select` system call and example code]: https://beej.us/guide/bgnet/html/split/slightly-advanced-techniques.html#select
[`select`]: https://www.man7.org/linux/man-pages/man2/select.2.html
[`netcat`]: https://www.man7.org/linux/man-pages/man1/ncat.1.html



debian/changelog
simple-chat (1.0.0) UNRELEASED; urgency=medium

  * Initial release.

 -- Benjamin Toll <ben@benjamintoll.com>  Sat, 03 Jun 2023 18:33:48 +0000


hash.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "hash.h"

/**
 * Add the new entry to the head of the linked list of the hash bucket.
 */
node_t *add_hash_entry(hash_table_t *hashtable, char* key, char *value) {
    if (hashtable == NULL) {
        fprintf(stderr, "No hash table\n");
        exit(1);
    }

    size_t hashval = hash(hashtable, key);
    node_t *new_node;

    if ((new_node = malloc(sizeof(node_t))) == NULL) {
        fprintf(stderr, "Could not allocate memory for new node\n");
        exit(2);
    }

    new_node->key = strdup(key);
    new_node->value = strdup(value);
    new_node->next = hashtable->table[hashval];

    hashtable->table[hashval] = new_node;
//     printf("%d\n", hashval);

    return new_node;
}

/**
 * Allocate memory for the hash table and its elements and initialize each element.
 */
hash_table_t *create_hashtable(size_t size) {
    if (size < 1) {
        fprintf(stderr, "Size must be greater than zero\n");
        exit(1);
    }

    size_t i;
    hash_table_t *new_table;

    if ((new_table = malloc(sizeof(hash_table_t))) == NULL) {
        fprintf(stderr, "Could not allocate memory for hash table\n");
        exit(2);
    }

    if ((new_table->table = malloc(sizeof(node_t) * size)) == NULL) {
        fprintf(stderr, "Could not allocate memory for hash table elements\n");
        exit(3);
    }

    for (i = 0; i < size; ++i)
        new_table->table[i] = NULL;

    new_table->size = size;

    return new_table;
}

void free_hashtable(hash_table_t *hashtable) {
    int size = hashtable->size, i;
    node_t *node, *tmp;

    for (i = 0; i < size; ++i) {
        // Remove the nodes in each linked list.
        node = hashtable->table[i];

        while (node) {
            node->key = NULL;
            node->value = NULL;

            tmp = node;
            node = node->next;
            free(tmp);
        }
    }

    free(hashtable->table);
    free(hashtable);
}

// TODO: Need a (much) better hashing algo! This is just to keep moving along.
size_t hash(hash_table_t *hashtable, char *key) {
    if (hashtable == NULL) {
        fprintf(stderr, "No hash table\n");
        exit(1);
    }

    size_t hashval = 0;

    for (; *key != '\0'; ++key)
        hashval = *key + (hashval << 5) - hashval;

    return hashval % hashtable->size;
}

node_t *lookup_hash_entry(hash_table_t *hashtable, char *key) {
    if (hashtable == NULL) {
        fprintf(stderr, "No hash table\n");
        exit(1);
    }

    size_t hashval = hash(hashtable, key);

    node_t *bucket = hashtable->table[hashval];

    if (bucket == NULL)
        return NULL;

    while (bucket && strncmp(bucket->key, key, strlen(key)) != 0)
        bucket = bucket->next;

    return bucket;
}



client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <netdb.h>

#define BUF_SIZE 4096
#define PORT "3333"

int main(int argc, char **argv) {
    struct addrinfo hints, *res, *p;
    int sock, r;
    char buf[BUF_SIZE];

    memset(&buf, 0, BUF_SIZE);
    memset(&res, 0, sizeof(hints));
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;

    if ((r = getaddrinfo("localhost", PORT, &hints, &res)) == -1) {
        fprintf(stderr, gai_strerror(r));
        exit(1);
    }

    for (p = res; p != NULL; p = p->ai_next) {
        if ((sock = socket(p->ai_family, p->ai_socktype, 0)) == -1) {
            perror("sock");
            continue;
        }

        if ((r = connect(sock, p->ai_addr, p->ai_addrlen)) == -1) {
            close(sock);
            perror("sock");
            continue;
        }

        break;
    }

    if (p == NULL) {
        fprintf(stderr, "client: could not establish connection\n");
        exit(2);
    }

    freeaddrinfo(res);

    for (;;) {
        int nread;

        while (printf("> "), fgets(buf, BUF_SIZE, stdin), !feof(stdin)) {
            if ((r = send(sock, buf, strlen(buf), 0)) == -1) {
                perror("send");
                exit(4);
            }

            if ((nread = recv(sock, buf, BUF_SIZE, 0)) < 0) {
                perror("recv");
                exit(5);
            } else {
                buf[nread] = '\0';
                printf("%s\n", buf);
            }
        }
    }

    return 0;
}



server.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netdb.h>
#include "hash.c"

#define BACKLOG 25
#define MAX_BUF_SIZE 4096
#define MIN_BUF_SIZE 4
#define NAME_SIZE 40
#define PORT "3333"

/*
void send_message(hash_table_t *hashtable, int sender_fd, int receiver_fd, char *buf, size_t nread) {
    // TODO
}
*/

/**
 * simple_chat [PORT]
 */
int main(int argc, char **argv) {
    int sock, maxfd, newfd, i, k;
    size_t r, j, nread, yes = 1, table_size = BACKLOG;

    struct addrinfo hints, *res, *p;
    struct sockaddr_storage client;

    char buf[MAX_BUF_SIZE], msg[MAX_BUF_SIZE], name[NAME_SIZE], fd_s[8];

    fd_set master, readfds;
    socklen_t sin_size;

    char *port = argc > 1 ?
        argv[1] :
        PORT;

    hash_table_t *hashtable = create_hashtable(table_size);

    memset(&fd_s, 0, strlen(fd_s));
    memset(&buf, 0, MAX_BUF_SIZE);
    memset(&msg, 0, MAX_BUF_SIZE);
    memset(&hints, 0, sizeof(hints));

    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags = AI_PASSIVE;

    if ((r = getaddrinfo(NULL, port, &hints, &res)) == -1) {
        perror(gai_strerror(r));
        exit(1);
    }

    for (p = res; p != NULL; p = p->ai_next) {
        if ((sock = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) {
            perror("socket");
            continue;
        }

        // Allow the socket to be reused.
        if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(int)) == -1) {
            close(sock);
            perror("setsockopt");
            continue;
        }

        if ((r = bind(sock, p->ai_addr, p->ai_addrlen)) == -1) {
            close(sock);
            perror("bind");
            continue;
        }

        break;
    }

    if (p == NULL) {
        fprintf(stderr, "server: could not connect\n");
        exit(2);
    }

    freeaddrinfo(res);

    if ((r = listen(sock, BACKLOG)) == -1) {
        perror("listen");
        exit(3);
    }

    fprintf(stderr, "chat server started on port %s...\n", port);

    maxfd = sock;
    FD_ZERO(&master);
    FD_ZERO(&readfds);
    FD_SET(sock, &master);

    for (;;) {
        readfds = master;

        if ((r = select(maxfd + 1, &readfds, NULL, NULL, NULL)) == -1) {
            perror("select");
            exit(4);
        }

        for (i = 0; i <= maxfd; ++i) {
            if (FD_ISSET(i, &readfds)) {
                if (i == sock) {
                    sin_size = sizeof(client);

                    if ((newfd = accept(i, (struct sockaddr *) &client, &sin_size)) == -1) {
                        perror("accept");
                        exit(5);
                    }

                    FD_SET(newfd, &master);
                    maxfd = newfd;

                    char greeting[] = "Hello, what is your name? ";
                    char msg[] = "Hi ";
                    char *newline;

                    send(newfd, greeting, strlen(greeting), 0);
                    nread = recv(newfd, name, NAME_SIZE, 0);

                    // The hashtable key for the user will be the stringified file descriptor.
                    sprintf(fd_s, "%d", newfd);

                    // Remove the newline.
                    newline = strchr(name, 13);
                    *newline = 0;

                    fprintf(stderr, "Received a new connection from %s (fd %d)\n", name, newfd);

                    add_hash_entry(hashtable, fd_s, name);

                    strncat(name, "\n", 1);
                    strncat(msg, name, nread);

                    send(newfd, msg, strlen(msg), 0);
                } else {
                    if ((nread = recv(i, buf, MAX_BUF_SIZE, 0)) < 0) {
                        perror("recv");
                        exit(6);
                    } else {
                        for (j = 0; j <= maxfd; ++j) {
                            // Don't send to either the server or the socket that wrote the message that was just received.
                            if (j > (sock + 1) && j != sock && j != i) {
                                // Get the sender's nickname.
                                sprintf(fd_s, "%d", i);
                                node_t *hash_entry;

                                if ((hash_entry = lookup_hash_entry(hashtable, fd_s)) != NULL) {
                                    // Handle Ctl-C.
                                    if (buf[0] == -1) {
                                        if ((snprintf(msg, strlen(hash_entry->value) + 20, "%s%s", hash_entry->value, " has left the chat\n")) == -1) {
                                            perror("snprintf");
                                            exit(6);
                                        }

                                        fprintf(stderr, "(fd %d) %s", i, msg);

                                        FD_CLR(i, &master);
                                        close(i);

                                        // Alert the other users that the user has left the chat.
                                        for (k = sock + 1; k <= maxfd; ++k)
                                            send(k, msg, strlen(msg), 0);
                                    } else {
                                        // Add the sender's nickname to the sender's chat message.
                                        if ((snprintf(msg, 4 + strlen(hash_entry->value) + nread, "%s%s%s %s", "<", hash_entry->value, ">", buf)) == -1) {
                                            perror("snprintf");
                                            exit(7);
                                        }

                                        send(j, msg, nread + strlen(msg), 0);
                                    }
//                                 send_message(hashtable, i, j, buf, nread);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    return 0;
}



hashtable.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "hashtable.h"

/**
 * Add the new entry to the head of the linked list of the hash bucket.
 */
node_t *add_hash_entry(hash_table_t *hashtable, char* key, char *value) {
    if (hashtable == NULL) {
        fprintf(stderr, "No hash table\n");
        exit(1);
    }

    size_t hashval = hash(hashtable, key);
    node_t *new_node;

    if ((new_node = (node_t *) malloc(sizeof(node_t))) == NULL) {
        fprintf(stderr, "Could not allocate memory for new node\n");
        exit(2);
    }

    new_node->key = strdup(key);
    new_node->value = strdup(value);
    new_node->next = hashtable->table[hashval];

    hashtable->table[hashval] = new_node;
//     printf("%d\n", hashval);

    return new_node;
}

/**
 * Allocate memory for the hash table and its elements and initialize each element.
 */
hash_table_t *create_hashtable(size_t size) {
    if (size < 1) {
        fprintf(stderr, "Size must be greater than zero\n");
        exit(1);
    }

    size_t i;
    hash_table_t *new_table;

    if ((new_table = (hash_table_t *) malloc(sizeof(hash_table_t))) == NULL) {
        fprintf(stderr, "Could not allocate memory for hash table\n");
        exit(2);
    }

    if ((new_table->table = (node_t **) malloc(sizeof(node_t) * size)) == NULL) {
        fprintf(stderr, "Could not allocate memory for hash table elements\n");
        exit(3);
    }

    for (i = 0; i < size; ++i)
        new_table->table[i] = NULL;

    new_table->size = size;

    return new_table;
}

void free_hashtable(hash_table_t *hashtable) {
    int size = hashtable->size, i;
    node_t *node, *tmp;

    for (i = 0; i < size; ++i) {
        // Remove the nodes in each linked list.
        node = hashtable->table[i];

        while (node) {
            node->key = NULL;
            node->value = NULL;

            tmp = node;
            node = node->next;
            free(tmp);
        }
    }

    free(hashtable->table);
    free(hashtable);
}

// TODO: Need a (much) better hashing algo! This is just to keep moving along.
size_t hash(hash_table_t *hashtable, char *key) {
    if (hashtable == NULL) {
        fprintf(stderr, "No hash table\n");
        exit(1);
    }

    size_t hashval = 0;

    for (; *key != '\0'; ++key)
        hashval = *key + (hashval << 5) - hashval;

    return hashval % hashtable->size;
}

node_t *lookup_hash_entry(hash_table_t *hashtable, char *key) {
    if (hashtable == NULL) {
        fprintf(stderr, "No hash table\n");
        exit(1);
    }

    size_t hashval = hash(hashtable, key);

    node_t *bucket = hashtable->table[hashval];

    if (bucket == NULL)
        return NULL;

    while (bucket && strncmp(bucket->key, key, strlen(key)) != 0)
        bucket = bucket->next;

    return bucket;
}



simple_chat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netdb.h>
#include "hashtable.c"
#include "simple_chat.h"

/**
 * simple_chat [PORT]
 */
int main(int argc, char **argv) {
//    int sock, maxfd, newfd, nread, i, j, k, r;
    int sock, maxfd, newfd, i, j, k, r;
    size_t yes = 1, table_size = BACKLOG;

    struct addrinfo hints, *res, *p;
    struct sockaddr_storage client;

    char fd_s[8];

    // fd_set is a struct that represents a set of file descriptors.
    // The max number is the value of the FD_SETSIZE macro (1024).
    fd_set master, readfds;
    socklen_t sin_size;

    char *port = argc > 1 ?
        argv[1] :
        PORT;

    hash_table_t *hashtable = create_hashtable(table_size);

    memset(&fd_s, 0, strlen(fd_s));
    memset(&hints, 0, sizeof(hints));

    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags = AI_PASSIVE;        // Fill in my IP for me.

    if ((r = getaddrinfo(NULL, port, &hints, &res)) == -1) {
        perror(gai_strerror(r));
        exit(1);
    }

    for (p = res; p != NULL; p = p->ai_next) {
        if ((sock = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) {
            perror("socket");
            continue;
        }

        // Sometimes, the bind call fails with "Address already in use."  A little bit of a socket that was
        // connected is still hanging around in the kernel, and it's hogging the port. You can either wait
        // for it to clear (a minute or so) or add code to your program allowing it to reuse the port.
        if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(int)) == -1) {
            close(sock);
            perror("setsockopt");
            continue;
        }

        if ((r = bind(sock, p->ai_addr, p->ai_addrlen)) == -1) {
            close(sock);
            perror("bind");
            continue;
        }

        break;
    }

    if (p == NULL) {
        fprintf(stderr, "server: could not connect\n");
        exit(2);
    }

    freeaddrinfo(res);

    if ((r = listen(sock, BACKLOG)) == -1) {
        perror("listen");
        exit(3);
    }

    fprintf(stderr, "Simple chat server started on port %s...\n", port);
    fprintf(stderr, "Clients can connect using telnet or netcat, i.e., `telnet|nc 127.0.0.1 %s`\n", port);

    maxfd = sock;
    // This macro removes / clears all the file descriptors from the set.  It
    // should be employed as the first step.
    FD_ZERO(&master);
    FD_ZERO(&readfds);
    // This macro adds a fd to a set.  If the fd is already preset, it's a no-op.
    FD_SET(sock, &master);

    for (;;) {
        readfds = master;

        // This blocks until a new client request or new data request is made.
        //
        // The first argument to select() is nfds, which should be set to the
        // highest-number fd in any of the three sets plus one.
        //
        // Next, the select() API defines three sets:
        //      1. readfds
        //          - watched to see if they're ready for reading
        //          - after select() has returned, readfds will be
        //            cleared of all fds except for those that are
        //            ready for reading
        //      2. writefds
        //          - watched to see if they're ready for writing
        //          - after select() has returned, writefds will be
        //            cleared of all fds except for those that are
        //            ready for writing
        //      3. exceptfds
        //          - watched for "exceptional conditions"
        //          - after select() has returned, exceptfds will be
        //            cleared of all fds except for those for which
        //            an exceptional condition has occurred
        // NULLing any of them restricts their use.  So, here we're only allowing
        // for readfds.
        //
        // Last, the timeout argument is a timeval structure.  NULL means
        // that select() will block indefinitely waiting for a fd to become
        // ready.  See the man page for more information.
        //
        // On success, select() returns the number of file descriptors contained
        // in the three returned descriptor sets (that is, the total number of bits
        // that are set in readfds, writefds, exceptfds).  The return value may be
        // zero if the timeout expired before any file descriptors became ready.
        if ((r = select(maxfd + 1, &readfds, NULL, NULL, NULL)) == -1) {
            perror("select");
            exit(4);
        }

        int nread;
        char buf[MAX_BUF_SIZE];
        memset(&buf, 0, MAX_BUF_SIZE);

        // If the master socket file descriptor is set, then a new request has been
        // made.  Recall that the kernel would redirect new requests to the master
        // socket file descriptor.
        //
        // This macro is used to test if the a fd is still present in a set (non-zero
        // if it's still present).
        // This must be checked continuously, since select() modified the contents
        // of the sets according to rules outlined in its man page.
        if FD_ISSET(sock, &readfds) {
            sin_size = sizeof(client);

            if ((newfd = accept(sock, (struct sockaddr *) &client, &sin_size)) == -1) {
                perror("accept");
                exit(5);
            }

            // Add new client communication socket file descriptor to the set.
            FD_SET(newfd, &master);

            if (newfd > maxfd)
                maxfd = newfd;

            char name[NAME_SIZE];
            memset(&name, 0, NAME_SIZE);
            char greeting[] = "Hello, what is your name? ";
            char *newline;

            send(newfd, greeting, strlen(greeting), 0);
            nread = recv(newfd, name, NAME_SIZE, 0);

            // The hashtable key for the user will be the stringified file descriptor.
            sprintf(fd_s, "%d", newfd);

            // Locate the substring of the newline (telnet) or newline & carriage return (netcat).
            if (!(newline = strstr(name, "\r\n"))) {
                newline = strstr(name, "\n");
            }

            // Copy the original string from up to the newline using pointer arithmetic.
            strncpy(buf, name, newline - name);
            buf[newline-name] = '\0';

            fprintf(stderr, "Received a new connection from %s (fd %d)\n", buf, newfd);

            add_hash_entry(hashtable, fd_s, buf);

            // Add a return character so the welcome message to the user isn't on the same
            // line as the first chat message.
            char initialmsg[40];
            sprintf(initialmsg, "Hi %s, welcome to the simple chat server!\n", buf);
            send(newfd, initialmsg, strlen(initialmsg), 0);
        } else // The communication file descriptor was activated.
        {
            for (i = 0; i <= maxfd; ++i) {
                if FD_ISSET(i, &readfds) {
                    // If nread is 0, the connection was closed.
                    if ((nread = recv(i, buf, MAX_BUF_SIZE, 0)) == -1) {
                        perror("recv");
                        exit(6);
                    }

                    // Don't send to either the server or the fd that wrote the message.
                    for (j = 0; j <= maxfd; ++j) {
                        if (j > sock && j != sock && j != i) {
                            // Get the sender's nickname.
                            sprintf(fd_s, "%d", i);
                            node_t *hash_entry;

                            if ((hash_entry = lookup_hash_entry(hashtable, fd_s)) != NULL) {
                                char msg[MAX_BUF_SIZE];
                                memset(&msg, 0, MAX_BUF_SIZE);

                                // Client closed the connection || Client pressed Ctl-C.
                                if (nread == 0 || buf[0] == -1) {
                                    if ((snprintf(msg, strlen(hash_entry->value) + 20, "%s%s", hash_entry->value, " has left the chat\n")) == -1) {
                                        perror("snprintf");
                                        exit(6);
                                    }

                                    fprintf(stderr, "(fd %d) %s", i, msg);

                                    // This macro removes the fd from the set.
                                    FD_CLR(i, &master);
                                    close(i);

                                    // Alert the other users that the user has left the chat.
                                    for (k = sock + 1; k <= maxfd; ++k)
                                        send(k, msg, strlen(msg), 0);

                                    // Break here to avoid duplicate "X has left the chat" messages.
                                    break;
                                } else {
                                    // Add the sender's nickname to the sender's chat message.
                                    memset(&msg, 0, MAX_BUF_SIZE);

                                    if ((snprintf(msg, 4 + strlen(hash_entry->value) + nread, "%s%s%s %s", "<", hash_entry->value, ">", buf)) == -1) {
                                        perror("snprintf");
                                        exit(7);
                                    }

                                    // If the return value of send doesn't match the amount of data sent,
                                    // it's up to us to send the remainder.
                                    // However, the amount of data that is anticipated to be sent using
                                    // this simple chat application shouldn't exceed 1KB a pop, which should
                                    // be able to bre sent in one go.
                                    send(j, msg, nread + strlen(msg), 0);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    free_hashtable(hashtable);
    return 0;
}

